### ISSUE TITLE
Controller and CMA-ES : number of parameters.

### ISSUE THREAD

**Issue by Caselles** [dylandjian/retro-contest-sonic#1]
Hey ! 

Thanks for the PyTorch code, it is pretty useful. The writeup is great too. 
 
I have two questions regarding CMA-ES and the Controller (which is the policy mapping states to actions). 

1) Regarding the number of parameters in the policy

The goal of CMA-ES is to optimize the policy of the controller, which is in your case the neural network defined [here](https://github.com/dylandjian/retro-contest-sonic/blob/master/models/controller.py). This neural network, composed of 2 FC layers, has over 1M parameters ((1024*2 + 200) * 512 + 512 * 4 = 1 153 024). Would you expect CMA-ES to work on such a high dimensional parameter space ? In the World Models paper, they justify using CMA-ES because they intentionally use a linear policy, which has less than 1k parameters. So it seem weird to use a MLP for the policy.

2) Regarding what is passed as input to CMA-ES

Also, I don't understand why the number of parameters passed to CMA-ES is `PARAMS_FC1 + LATENT_VEC + 512`. Isn't this number should be the number of parameters in the policy aka Controller ? Then it should be `(PARAMS_FC1 + LATENT_VEC) * 512 + 512 * ACTION_SPACE` (as in the calculation mentioned before).

**Comment by Caselles** [comment]
Regarding question 2., I think I have spotted the problem. 

[Here](https://github.com/dylandjian/retro-contest-sonic/blob/master/train_controller.py#L39) you are copying the parameters sampled (and optimized) by CMA-ES into the policy. 

```python
new_w1 = torch.tensor(solution[0:PARAMS_FC1 + LATENT_VEC],\
                                dtype=torch.double, device=DEVICE)
new_w2 = torch.tensor(solution[PARAMS_FC1 + LATENT_VEC:],\
                                dtype=torch.double, device=DEVICE)
```

But I think you misunderstood the number of parameters in the FC layer. A FC layer of input size `n` and output size `m` has `n * m` parameters. [Here is a way to see it.](https://stats.stackexchange.com/questions/256342/how-many-learnable-parameters-does-a-fully-connected-layer-have-without-the-bias) But you are only copying `n` parameters as if there are only `n` parameters in the layer. But you don't get any error from PyTorch because it automatically decides to copy the `n` parameters in every `m` rows of the layer. Here is a minimal example of this behavior:  

![image](https://user-images.githubusercontent.com/19774802/42125750-c2ba6918-7c7c-11e8-8794-0e2759840520.png)

If you correct this then the implementation should be correct.

Regarding question 1. , I still believe that CMA-ES would work much better if the policy has less parameters, around a few hundred (which is what is advised here : https://pypi.org/project/cma/) rather than roughly 1M.

Let me know what if you think I made any mistakes :)

**Comment by dylandjian** [comment]
Yes you are right indeed !

In fact, I tried to fiddle around with the number of parameters just to see if it did work better than with just a few parameters (which should not), and actually completely forgot that I had implemented that.
My mistake ! 

I will correct it soon, thanks for pointing that out ! 👍

**Comment by dylandjian** [comment]
This was indeed a really big error ! 

Now there is *only* 8992 (1024 hidden parameters in the LSTM * 2 because we use both the cell state and the hidden state + 200 for the latent dimension of the VAE * 4 because of the number of actions) in the Controller model, which leads to way better average scores.    

In only 5 generations, Sonic is already showing good behavior (going right and jumping when stuck), and the variance in the score of each individual controller is much lower than before ! This seems to actually be converging, just like in the paper. I will maybe try to retrain it if I get access to more compute than what I have at home.

Nice spot, closing the issue which seems to be fixed in the last commit (#33)
